File: src/main/java/model/wdkugdn/Wdkugdn.java
package model.wdkugdn;

public class Wdkugdn {

    public static Double[][] getMap() {
        Double[][] map = new Double[][]{
                {5.0480, 12.8367, 18.1960, 22.4017, 28.3302, 31.8934, 35.5894, 38.4033, 41.3622, 43.8038, 48.1041, 50.2039}
        };

        return map;
    }

    public static Double[] getXAxis() {
        return new Double[]{240d, 520d, 760d, 1000d, 1520d, 2000d, 2520d, 3000d, 3520d, 4000d, 5000d, 6000d};
    }

    public static Double[] getYAxis() {
        return new Double[]{0d};
    }


}

File: src/main/kotlin/domain/model/wdkugdn/Wdkugdn.kt
package domain.model.wdkugdn

import domain.math.LinearInterpolation
import domain.math.map.Map3d
import domain.model.load.EngineLoad
import java.util.Arrays
import kotlin.math.abs

object Wdkugdn {
    fun calculateWdkugdn(wdkugdn: Map3d, kfwdkmsn: Map3d, displacement: Double): Map3d {
        val xAxis = wdkugdn.xAxis
        val correctedWdkugdn = Map3d(wdkugdn)

        for (i in xAxis.indices) {
            val rpm = xAxis[i]
            val chokedAirflow = EngineLoad.getAirflow(1.0, rpm, displacement) * 3.6 / 0.528

            var rpmIndex = Arrays.binarySearch(kfwdkmsn.xAxis, rpm)
            if (rpmIndex < 0) rpmIndex = abs(rpmIndex + 1)
            rpmIndex = minOf(rpmIndex, kfwdkmsn.xAxis.size - 1)

            val throttleAngle = DoubleArray(kfwdkmsn.yAxis.size) { j -> kfwdkmsn.zAxis[j][rpmIndex] }

            var airflowIndex = Arrays.binarySearch(kfwdkmsn.yAxis, chokedAirflow)
            if (airflowIndex < 0) airflowIndex = abs(airflowIndex + 1)
            airflowIndex = minOf(airflowIndex, kfwdkmsn.yAxis.size - 1)
            airflowIndex = maxOf(airflowIndex, 1)

            val x = arrayOf(kfwdkmsn.yAxis[airflowIndex - 1], kfwdkmsn.yAxis[airflowIndex])
            val y = arrayOf(throttleAngle[airflowIndex - 1], throttleAngle[airflowIndex])
            val xi = arrayOf(chokedAirflow)

            val chokedThrottleAngle = LinearInterpolation.interpolate(x, y, xi)
            correctedWdkugdn.zAxis[0][i] = chokedThrottleAngle[0]
        }

        return correctedWdkugdn
    }
}
